wagtail/search/tests/test_queries.py

```diff
@@ -221,3 +222,115 @@ def test_two_filters_and_query(self):
+        self.assertEqual(
+            balanced_reduce(add, [['a', 'c'], [], ['d', 'w']], []),
+            ['a', 'c', 'd', 'w']
+        )
+        self.assertEqual(balanced_reduce(lambda x, y: x * y, range(2, 8), 1), 5040)
+        self.assertEqual(
+            balanced_reduce(lambda x, y: x * y, range(2, 21), 1),
+            2432902008176640000
+        )
+        self.assertEqual(balanced_reduce(add, Squares(10)), 285)
+        self.assertEqual(balanced_reduce(add, Squares(10), 0), 285)
+        self.assertEqual(balanced_reduce(add, Squares(0), 0), 0)
+        self.assertRaises(TypeError, balanced_reduce)
+        self.assertRaises(TypeError, balanced_reduce, 42, 42)
+        self.assertRaises(TypeError, balanced_reduce, 42, 42, 42)
+        self.assertEqual(balanced_reduce(42, "1"), "1")  # func is never called with one item
+        self.assertEqual(balanced_reduce(42, "", "1"), "1")  # func is never called with one item
+        self.assertRaises(TypeError, balanced_reduce, 42, (42, 42))
+        self.assertRaises(TypeError, balanced_reduce, add, [])  # arg 2 must not be empty sequence with no initial value
+        self.assertRaises(TypeError, balanced_reduce, add, "")
+        self.assertRaises(TypeError, balanced_reduce, add, ())
+        self.assertRaises(TypeError, balanced_reduce, add, object())
+
+        class TestFailingIter:
+            def __iter__(self):
+                raise RuntimeError
+
+        self.assertRaises(RuntimeError, balanced_reduce, add, TestFailingIter())
+
+        self.assertEqual(balanced_reduce(add, [], None), None)
+        self.assertEqual(balanced_reduce(add, [], 42), 42)
+
+        class BadSeq:
+            def __getitem__(self, index):
+                raise ValueError
+
+        self.assertRaises(ValueError, balanced_reduce, 42, BadSeq())
+
+    # Test reduce()'s use of iterators.
+    def test_iterator_usage(self):
+        class SequenceClass:
+            def __init__(self, n):
+                self.n = n
+
+            def __getitem__(self, i):
+                if 0 <= i < self.n:
+                    return i
+                else:
+                    raise IndexError
+
+        from operator import add
+        self.assertEqual(balanced_reduce(add, SequenceClass(5)), 10)
+        self.assertEqual(balanced_reduce(add, SequenceClass(5), 42), 52)
+        self.assertRaises(TypeError, balanced_reduce, add, SequenceClass(0))
+        self.assertEqual(balanced_reduce(add, SequenceClass(0), 42), 42)
+        self.assertEqual(balanced_reduce(add, SequenceClass(1)), 0)
+        self.assertEqual(balanced_reduce(add, SequenceClass(1), 42), 42)
+
+        d = {"one": 1, "two": 2, "three": 3}
+        self.assertEqual(balanced_reduce(add, d), "".join(d.keys()))
+
+    # This test is specific to balanced_reduce
+    def test_is_balanced(self):
+        # Tests that balanced_reduce returns the object as a balanced tree
+        class Node:
```

Comment:
Just realised this class is completely unnecessary. I could just use the letter strings as the nodes... :facepalm: